### ISSUE TITLE
`--capture-movie` exits with `Movie capture error` over PTP/IP (`--capture-image` also errors) - Canon EOS R

### ISSUE THREAD

**Issue by chrishobcroft** [gphoto/gphoto2#356]
**Describe the bug**

When connecting to a Canon EOS R using PTP/IP over WiFi, the `--capture-movie` and `--capture-image` commands both give errors.

`debug.log` files are attached below.

**Name the camera**

Canon EOS R

```
ubuntu@cibulka:~$ gphoto2 --port ptpip:192.168.1.146 --summary
Camera summary:                                                                
Manufacturer: Canon.Inc
Model: Canon EOS R
  Version: 3-1.7.0
  Serial Number: *****************************70
Vendor Extension ID: 0xb (1.0)

Capture Formats: JPEG
Display Formats: Association/Directory, Script, DPOF, MS AVI, MS Wave, JPEG, CRW, Unknown(b103), Unknown(bf02), Defined Type, Unknown(b104), Unknown(b105)

Device Capabilities:
	File Download, File Deletion, File Upload
	No Image Capture, No Open Capture, Canon EOS Capture, Canon EOS Capture 2
	Canon Wifi support

Storage Devices Summary:
store_00020001:
	StorageDescription: SD
	VolumeLabel: None
	Storage Type: Removable RAM (memory card)
	Filesystemtype: Digital Camera Layout (DCIM)
	Access Capability: Read-Write
	Maximum Capability: 127831900160 (121910 MB)
	Free Space (Bytes): 113060872192 (107823 MB)
	Free Space (Images): -1

Device Property Summary:
Property 0xd402:(read only) (type=0xffff) 'Canon EOS R'
Property 0xd407:(read only) (type=0x6) 1
Property 0xd406:(readwrite) (type=0xffff) 'Unknown Initiator'
Property 0xd303:(read only) (type=0x2) 1
Battery Level(0x5001):(read only) (type=0x2) Enumeration [100,0,75,0,50] value: 91% (91)
```

**libgphoto2 and gphoto2 version**
```
ubuntu@cibulka:~$ gphoto2 --version
gphoto2 2.5.23

Copyright (c) 2000-2019 Marcus Meissner and others

gphoto2 comes with NO WARRANTY, to the extent permitted by law. You may
redistribute copies of gphoto2 under the terms of the GNU General Public
License. For more information about these matters, see the files named COPYING.

This version of gphoto2 is using the following software versions and options:
gphoto2         2.5.23         gcc, popt(m), exif, cdk, aa, jpeg, readline
libgphoto2      2.5.25         standard camlibs (SKIPPING lumix), gcc, ltdl, EXIF
libgphoto2_port 0.12.0         iolibs: disk ptpip serial usb1 usbdiskdirect usbscsi, gcc, ltdl, EXIF, USB, serial without locking
```

**To Reproduce**

1. Turn on camera
2. Go to WiFi settings, and select "Connect to smartphone"
3. Connect camera to same Access Point as computer
4. Run the following command on the computer:
```
gphoto2 --port ptpip:192.168.1.146 --summary
```
5. Press OK on the camera to allow the incoming connection
6. Observe a summary is given.
7. Also, run the following, with successful responses from the camera (see below for full responses):
```
gphoto2 --port ptpip:192.168.1.146 --list-files
gphoto2 --port ptpip:192.168.1.146 --storage-info
gphoto2 --port ptpip:192.168.1.146 --abilities
```
8. Run
```
gphoto2 --port ptpip:192.168.1.146 --capture-movie
```
9. Observe the camera's lights reacting to the request, but `gphoto2` giving the error:
```
Capturing preview frames as movie to 'movie.mjpg'. Press Ctrl-C to abort.      
ERROR: Movie capture error... Exiting.
Movie capture finished (0 frames)
```
Here's the [debug.log](https://github.com/gphoto/gphoto2/files/5048265/debug.log)

10. Incidentally, `--capture-image` also appears to make the camera capture and image, but `gphoto2` fails with
```
ERROR: Could not capture image.                                                
ERROR: Could not capture.
```
Here's the [debug.log](https://github.com/gphoto/gphoto2/files/5048281/debug.log)

**Evidence of success with other `gphoto2` commands**
```
ubuntu@cibulka:~$ gphoto2 --port ptpip:192.168.1.146 --abilities
Abilities for camera             : PTP/IP Camera                               
Serial port support              : no
USB support                      : no
Capture choices                  :
                                 : Image
                                 : Preview
Configuration support            : yes
Delete selected files on camera  : yes
Delete all files on camera       : no
File preview (thumbnail) support : yes
File upload support              : yes
```
```
ubuntu@cibulka:~$ gphoto2 --port ptpip:192.168.1.146 --list-files
There is no file in folder '/'.                                                
There is no file in folder '/store_00020001'.
There is no file in folder '/store_00020001/DCIM'.
There are 48 files in folder '/store_00020001/DCIM/100EOS_R'.
#1     2R3A0003.MP4               rd 64347 KB video/mp4
#2     2R3A0006.JPG               rd  8397 KB image/jpeg
#3     2R3A0011.JPG               rd  8350 KB image/jpeg
#4     2R3A0013.MP4               rd 457241 KB video/mp4
#5     2R3A0014.MP4               rd 1324272 KB video/mp4
#6     2R3A0015.MP4               rd 1356702 KB video/mp4
#7     2R3A0016.MP4               rd 1791560 KB video/mp4
#8     2R3A0017.MP4               rd 2004371 KB video/mp4
#9     2R3A0018.MP4               rd 394105 KB video/mp4
#10    2R3A0019.JPG               rd  5014 KB image/jpeg
#11    2R3A0020.JPG               rd  4909 KB image/jpeg
#12    2R3A0022.JPG               rd  6648 KB image/jpeg
#13    2R3A0024.JPG               rd  5747 KB image/jpeg
#14    2R3A0025.JPG               rd  6121 KB image/jpeg
#15    2R3A0028.MP4               rd 1056230 KB video/mp4
#16    2R3A0029.MP4               rd 325415 KB video/mp4
#17    2R3A0030.MP4               rd 379830 KB video/mp4
#18    2R3A0031.MP4               rd 93235 KB video/mp4
#19    2R3A0032.MP4               rd 69361 KB video/mp4
#20    2R3A0040.MP4               rd 91842 KB video/mp4
#21    2R3A0041.MP4               rd 82138 KB video/mp4
#22    2R3A0042.MP4               rd 70225 KB video/mp4
#23    2R3A0043.MP4               rd 119416 KB video/mp4
#24    2R3A0044.MP4               rd 110668 KB video/mp4
#25    2R3A0045.MP4               rd 96486 KB video/mp4
#26    2R3A0046.MP4               rd 414274 KB video/mp4
#27    2R3A0047.MP4               rd 122742 KB video/mp4
#28    2R3A0048.MP4               rd 687171 KB video/mp4
#29    2R3A0049.MP4               rd 2391532 KB video/mp4
#30    2R3A0050.JPG               rd 10791 KB image/jpeg
#31    2R3A0052.JPG               rd  6778 KB image/jpeg
#32    2R3A0053.JPG               rd  9720 KB image/jpeg
#33    2R3A0054.JPG               rd  9217 KB image/jpeg
#34    2R3A0056.JPG               rd  7270 KB image/jpeg
#35    2R3A0057.JPG               rd  9931 KB image/jpeg
#36    2R3A0058.JPG               rd 11078 KB image/jpeg
#37    2R3A0060.MP4               rd 17157 KB video/mp4
#38    2R3A0065.JPG               rd  6742 KB image/jpeg
#39    2R3A0066.JPG               rd  8304 KB image/jpeg
#40    2R3A0067.JPG               rd  9041 KB image/jpeg
#41    2R3A0071.MP4               rd 15929 KB video/mp4
#42    2R3A0072.MP4               rd 117712 KB video/mp4
#43    2R3A0077.JPG               rd  7523 KB image/jpeg
#44    2R3A0078.JPG               rd  7179 KB image/jpeg
#45    2R3A0079.JPG               rd  7377 KB image/jpeg
#46    2R3A0080.MP4               rd 81528 KB video/mp4
#47    2R3A0081.JPG               rd  5140 KB image/jpeg
#48    2R3A0082.MP4               rd 326044 KB video/mp4
There is no file in folder '/store_00020001/MISC'.
```
```
ubuntu@cibulka:~$ gphoto2 --port ptpip:192.168.1.146 --storage-info
[Storage 0]                                                                    
description=SD
basedir=/store_00020001
access=0 Read-Write
type=4 Removable RAM
fstype=4 Camera layout (DCIM)
totalcapacity=124835840 KB
free=110411008 KB
```

**Comment by msmeissn** [comment]
Not seeing any problem so far in the logs ... for the record, the same commands work over USB?

**Comment by chrishobcroft** [comment]
I can confirm, that running the command, _with the camera connected by USB_, successfully pulls a movie from the camera, and pipes it into a `ffmpeg` command, which presents a webcam on `/dev/video2` (using `v4l2loopback`) for use as a webcam in OBS (for example):

![image](https://user-images.githubusercontent.com/2212651/90626299-63491000-e238-11ea-90cc-d1dffbbc643b.png)

This is the command being run:

```
gphoto2 --stdout --capture-movie | ffmpeg -i - -vcodec rawvideo -pix_fmt yuv420p -threads 0 -f v4l2 /dev/video2
```

Running the same command, but this time using `ptpip` over WiFi (below), yields the following:

![image](https://user-images.githubusercontent.com/2212651/90625981-f9306b00-e237-11ea-91c7-5ae60540049a.png)

This is the command being run:

```
gphoto2 --port ptpip:192.168.43.117 --stdout --capture-movie | ffmpeg -i - -vcodec rawvideo -pix_fmt yuv420p -threads 0 -f v4l2 /dev/video2
```

**Comment by chrishobcroft** [comment]
Looking at [the debug log from running `--capture-movie`](https://github.com/gphoto/gphoto2/files/5048265/debug.log), there are a few possible clues, something about `get_viewfinder_image failed`:
```
3.990175 ptp_ptpip_getresp           (2): PTPIP_CMD_RESPONSE
3.990182 camera_capture_preview [library.c:3068](0): 'ret' failed: get_viewfinder_image failed (0xa102: Camera Not Ready)
3.990187 gp_camera_capture_preview [gphoto2-camera.c:1403](0): 'camera->functions->capture_preview ( camera, file, context)' failed: -1
3.990563 gp_camera_free              (2): Freeing camera...
3.990574 gp_camera_exit              (2): Exiting camera ('PTP/IP Camera')...
3.990581 ptp_ptpip_sendreq           (2): Sending PTP_OC 0x9116 (EOS_GetEvent) request...
3.990643 ptp_ptpip_sendreq           (3): ptpip/oprequest data: (hexdump of 18 bytes)
0000  12 00 00 00 06 00 00 00-01 00 00 00 16 91 40 00  ..............@.
0010  00 00                  -                         ..              
```
and then this gets reported ~150 times:
```
3.995917 camera_exit                 (2): missed EOS ptp type 0
```

**Comment by chrishobcroft** [comment]
So, apparently, Canon allow you to activate a feature on CCAPI that allows pulling of video source over wifi. Trying now.

**Comment by chrishobcroft** [comment]
Revisiting this after ~6 years, on a **Canon EOS R5** (firmware 3-2.2.0) over PTP/IP. I found the cause of the `Movie capture error`, and it turns out to be a one-value config fix — plus two other findings worth recording.

## TL;DR — live view over PTP/IP works if you set the EVF output to **TFT + PC**

```sh
gphoto2 --port ptpip:<camera-ip> \
  --set-config output=3 \
  --set-config viewfinder=1 \
  --stdout --capture-movie | ffplay -f mjpeg -i -
```

`output` is the `0xd1b0` **`EOS_EVFOutputDevice`** property:

```
Choice: 0 Off
Choice: 1 TFT      <- default over PTP/IP; live view fails
Choice: 2 PC       <- ALSO fails (0xa102)
Choice: 3 TFT + PC <- works
```

Over PTP/IP the camera comes up as `output=TFT`, i.e. live view is routed **only to the camera's own screen**, so `EOS_GetViewFinderData` correctly answers `0xa102 Camera Not Ready`. That's exactly the line from my original 2020 log:

```
camera_capture_preview: 'ret' failed: get_viewfinder_image failed (0xa102: Camera Not Ready)
```

The non-obvious part: setting `output=2` (**PC alone**) is *not* enough — the R5 still returns `0xa102`. It appears to refuse to blank its own screen over PTP/IP. Only `output=3` (**TFT + PC**) works. I confirmed the property is genuinely applied (`event: prop d1b0 value changed (EOS_EVFOutputDevice)`, readback `Current: PC`) and it *still* failed on `2` — so this isn't a set-config failure.

Verified working: **41 frames** via `--capture-movie=40`, and continuous capture rendering live in `ffplay` (`mjpeg 960x640`), no cable.

Tested on an R5; I no longer have the EOS R from the original report, but given the shared EOS PTP layer it's worth others trying `output=3` there too.

## Finding 2 — the PTP/IP pairing handshake needs an on-camera confirmation, and gphoto2 times out before you can give it

On the R5, the first connection to a not-yet-paired host looks like a protocol failure:

```
read PTPIPHeader: Connection timed out
*** Error (-7: 'I/O problem') ***
```

What's actually happening: gphoto2's `PTPIP_INIT_COMMAND_REQUEST` **is** received, and the camera puts up a prompt — *"PC &lt;hostname&gt; found. Connect to the PC."* — and holds the connection while waiting for the user to press **OK on the camera body**. gphoto2's socket read timeout expires long before that, so it closes the socket and reports a timeout.

Notes for anyone hitting this:

* Press **OK on the camera** even after gphoto2 has given up — the pairing still registers. The camera may then show **`Err 11` ("Connection target not found")**, which is harmless: it's just the camera failing to find the PC whose socket we already closed.
* Afterwards, `gphoto2 --port ptpip:<ip> --summary` connects **instantly**, no prompt.
* The camera grants roughly **one init attempt per arming** of the EOS-Utility pairing screen. Every subsequent connection gets a TCP **RST** until it's re-armed — so retry loops (and even `nc -z` probes) silently burn the one slot you have and make this look far more broken than it is.

I confirmed the handshake by holding the socket open from a small Python client long enough to press OK, and capturing the real ack:

```
REPLY len=58 type=2 (INIT_COMMAND_ACK)
camera name: EOSR5_7D4F96
```

A longer (or configurable) init read timeout would make this behave much better out of the box.

## Finding 3 — live view over PTP/IP is throttled by the camera to ~1 Mbit/s

Measured on the R5, same camera, same session settings, only `liveviewsize` changed:

| `liveviewsize` | frame size | fps | session length |
|---|---|---|---|
| Medium (default) | ~125 KB | **0.7** | ~32 s |
| Small | ~37 KB | **2.93** | ~59 s |

Both land at **≈0.7–0.86 Mbit/s**, so frame rate is simply that budget divided by frame size. There is no size below `Small`, so **~3 fps is the ceiling**. The same camera does **25 fps over USB**.

This is **not** network or client-side:

* `EOS_GetViewFinderData` (`0x9153`) blocks **~950 ms** per frame, while `EOS_GetEvent` (`0x9116`) round-trips in **22 ms** on the **same socket**, moments apart.
* gphoto2's capture loop has **no sleeps** — it's blocked waiting on the camera.
* Link was 300 Mbit/s, 0% packet loss, ~5 ms RTT.

```
3.322925  Sending  EOS_GetViewFinderData
4.274561  Reading  response              <- 950 ms
4.274727  Sending  EOS_GetEvent
4.296783  response                       <- 22 ms
```

(For context: Canon's own Camera Connect app gets smooth, near-realtime video from the same camera — but over a BLE-bootstrapped **Wi-Fi Direct** link with a proprietary protocol, so it isn't a like-for-like transport comparison. It does show the hardware is capable of far more.)

## Finding 4 — `--capture-movie` aborts on a single `PTP Timeout` instead of retrying (possible easy fix)

Sessions reliably die after ~30–60 s with:

```
get_viewfinder_image failed (0x02fa: PTP Timeout)
'camera->functions->capture_preview (camera, file, context)' failed: -1
ERROR: Movie capture error... Exiting.
```

The preceding requests all return `0x2001 OK`; a single stalled `EOS_GetViewFinderData` ends the whole capture. Given each request already takes ~1 s over PTP/IP, occasionally exceeding the timeout seems inevitable — **retrying that request instead of aborting would make sessions dramatically more robust**. Right now the only workaround is an external supervisor that restarts `--capture-movie` (and the R5 dislikes rapid reconnects, so it needs a ~15 s backoff).

## Environment

* Canon EOS R5, firmware `3-2.2.0`
* `gphoto2` 2.5.28, `libgphoto2` 2.5.31, `libgphoto2_port` 0.12.2 (`ptpip` iolib)
* Camera in EOS Utility connection mode, over WiFi (infrastructure and camera-AP mode both reach the same conclusions)
